### PR TITLE
cli: Handle legitimate provider exec errors

### PR DESCRIPTION
**Pull request by cmatsuoka**
When a build provider returns an error for a build that failed for
legitimate reasons, handle this on the outer error handler and don't
raise an exception.

Fixes SNAPCRAFT-FK
Fixes SNAPCRAFT-EF
Fixes SNAPCRAFT-E5

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: tests/spread/multipass/error-handling/snaps/exit1/snap/snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+name: exit1
+base: core18
```
Comment (sergiusens) [inline tests/spread/multipass/error-handling/snaps/exit1/snap/snapcraft.yaml]: if using the set-base script in task.yaml, either remove this line here or add more logic to the script (the former would do for now)

File: spread.yaml [hunk 2]
```diff
@@ -155,6 +155,12 @@ suites:
  tests/spread/general/:
    summary: tests of snapcraft core functionality
 
+ tests/spread/multipass/:
+   summary: tests of snapcraft using multipass
+   # Keep this 18.04 only for now as it is the only stable base.
+   systems: [ubuntu-18.04*]
+   manual: true
```
Comment (sergiusens) [inline spread.yaml]: let's try and remove this line to see if the tests actually run.

**Comment by sergiusens** [comment]
Some more unit tests would be good here, as this is a delicate component (the exit handler logic)

File: tests/spread/multipass/error-handling/snaps/exit1/snap/snapcraft.yaml (hunk 1 above)
Comment (cmatsuoka) [reply tests/spread/multipass/error-handling/snaps/exit1/snap/snapcraft.yaml]: Line removed.

File: spread.yaml (hunk 2 above)
Comment (cmatsuoka) [reply spread.yaml]: Line removed.

File: tests/unit/cli/test_errors.py [hunk 3]
```diff
@@ -161,7 +161,7 @@ def test_provider_error_outer(self, isfile_function):
             fixtures.EnvironmentVariable("SNAPCRAFT_BUILD_ENVIRONMENT", "host")
         )
         self._raise_exec_error()
-        self.move_mock.assert_called_once()
+        self.assertTrue(self.move_mock.call_count == 1)
```
Comment (sergiusens) [inline tests/unit/cli/test_errors.py]: `self.assertThat(self.move_mock.call_count, Equals(1))`

File: tests/unit/cli/test_errors.py [hunk 4]
```diff
@@ -172,7 +172,7 @@ def test_provider_error_host(self, isfile_function):
         )
         self._raise_other_error()
         self.move_mock.assert_not_called()
-        self.traceback_mock.assert_called_once()
+        self.assertTrue(self.traceback_mock.call_count == 1)
```
Comment (sergiusens) [inline tests/unit/cli/test_errors.py]: `self.assertThat(self.move_mock.call_count, Equals(1))`

File: tests/unit/cli/test_errors.py (hunk 3 above)
Comment (sergiusens) [reply tests/unit/cli/test_errors.py]:
```suggestion
        self.assertThat(self.move_mock.call_count, Equals(1))
```

**Pull request merged**